### PR TITLE
feat: add `amendments/vote` endpoint

### PR DESCRIPTION
**Pull request by pdp2121**
## High Level Overview of Change

Add an api endpoint to retrieve lists of enabled and voting amendments for a network.
Also, add functionality to manually edit amendments from db.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Test Plan

Test on staging for a few days.

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -24,7 +24,9 @@ There are 3 folders in `src`, corresponding to the 3 processes that the VHS runs
   * `/network/validator/:publicKey/manifests`: Returns the manifests of a specific validator.
   * `/network/validator/:publicKey/reports`: Returns more detailed information about the reliability of a specific validator.
   * `/network/amendments/info`: Returns general information about known amendments.
-  * `/network/amendments/info/param`: Returns general information about a specific amendment by name or ID.
+  * `/network/amendments/vote/:network`: Returns list of enabled and in voting amendments on a specific network.
+  * `/network/amendment/info/param`: Returns general information about a specific amendment by name or ID.
```
Comment (ckniffen) [inline ARCHITECTURE.md:28]:
```suggestion
  * `/network/amendment/info/:nameOrID`: Returns general information about a specific amendment by name or ID.
```

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -24,7 +24,9 @@ There are 3 folders in `src`, corresponding to the 3 processes that the VHS runs
...
+  * `/network/amendment/vote/:network/:identifier`: Returns the voting status of a specific amendment by name or ID.
```
Comment (ckniffen) [inline ARCHITECTURE.md:29]:
```suggestion
  * `/network/amendment/vote/:network/:nameOrID`: Returns the voting status of a specific amendment by name or ID.
```

File: src/connection-manager/data/amendments_enabled.json [hunk 3]
```diff
@@ -309,112 +309,6 @@
     }
     ]
   },
-  {"networks": "dev",
```
Comment (ckniffen) [inline src/connection-manager/data/amendments_enabled.json:312]: Why was this removed?

File: src/api/routes/v1/amendments.ts [hunk 4]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
 const log = logger({ name: 'api-amendments' })
 
-const cache: Cache = {
+const cacheInfo: CacheInfo = {
   amendments: [],
   time: Date.now(),
 }
 
+const cacheVote: CacheVote = {
+  networks: new Map<string, Array<AmendmentsEnabled | AmendmentInVoting>>(),
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:84]: Should this be ` Map<number, Array<AmendmentsEnabled | AmendmentInVoting>>()`?

File: src/api/routes/v1/amendments.ts [hunk 5]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
-    cache.amendments.sort((prev: AmendmentsInfo, next: AmendmentsInfo) => {
-      if (isEarlierVersion(prev.rippled_version, next.rippled_version)) {
-        return 1
+    cacheInfo.amendments = await query('amendments_info').select('*')
+    cacheInfo.amendments.sort((prev: AmendmentsInfo, next: AmendmentsInfo) =>
+      sortByVersion(prev, next),
+    )
+    cacheInfo.time = Date.now()
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TODO: clean up
+  } catch (err: any) {
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:121]: Use `unknown` instead of `any`.

File: src/api/routes/v1/amendments.ts [hunk 6]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
+    (validator) => validator.unl,
+  ).length
+  const dbUNL = await query('validators')
+    .count('signing_key AS count')
+    .whereNotNull('unl')
+    .andWhere('chain', network_id)
+
+  const totalUnl: number = dbUNL[0].count
+
+  // eslint-disable-next-line require-atomic-updates -- no race condition.
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:217]: Can you explain why this fails atomic updates?

File: src/api/routes/v1/amendments.ts [hunk 7]
```diff
@@ -108,3 +365,77 @@ export async function handleAmendmentInfo(
...
+      // eslint-disable-next-line max-depth -- Disable for this function.
+      if (amendment.length > 0) {
+        res.send({
+          result: 'success',
+          amendment: amendment[0],
+        })
+      } else {
+        res
+          .status(404)
+          .send({ result: 'error', message: "incorrect amendment's id/name" })
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:433]:
```suggestion
          .send({ result: 'error', message: "amendment with id/name not found" })
```

File: src/api/routes/v1/amendments.ts [hunk 8]
```diff
@@ -108,3 +365,77 @@ export async function handleAmendmentInfo(
...
+): Promise<void> {
+  try {
+    const { network, identifier } = req.params
+    if (Date.now() - cacheVote.time > 60 * 1000) {
+      await cacheAmendmentsVote()
+    }
+    const networkVotes:
+      | Array<AmendmentsEnabled | AmendmentInVoting>
+      | undefined = cacheVote.networks.get(network)
+    if (networkVotes) {
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:420]: Teverse this if statement. That will hep reduce nesting.

File: src/api/routes/v1/validator.ts [hunk 9]
```diff
@@ -245,8 +245,13 @@ export async function handleValidator(
     }
 
     const public_key = req.params.publicKey
-    let validator: ValidatorResponse | undefined = cache.validators.find(
-      (resp: ValidatorResponse) => resp.validation_public_key === public_key,
+    let validator: ValidatorResponse | undefined = JSON.parse(
+      JSON.stringify(
```
Comment (ckniffen) [inline src/api/routes/v1/validator.ts:249]: Why do we parse stringify just to parse again?

File: src/connection-manager/connections.ts [hunk 10]
```diff
@@ -156,18 +182,15 @@ async function handleWsMessageSubscribeTypes(
...
-        reserve_base: current_ledger.reserve_base,
-        reserve_inc: current_ledger.reserve_inc,
-      }
-      network_fee.set(networks, fee)
-    }
-    if (ledger_hashes.length > LEDGER_HASHES_SIZE) {
-      ledger_hashes.shift()
+    void handleLedger(data as StreamLedger, ledger_hashes, networks)
+    // eslint-disable-next-line no-prototype-builtins -- Safeguards against some strange streams data.
+  } else if (!data.hasOwnProperty('id')) {
```
Comment (ckniffen) [inline src/connection-manager/connections.ts:187]:
```suggestion
  } else if (Object.prototype.hasOwnProperty.call(data, 'id')) {
```

This will allow you to get rid of the linter exception.

File: src/api/routes/v1/amendments.ts [hunk 11]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
+    name: string
+    rippled_version: string | undefined | null
+    threshold: string
+    consensus: string
+    validators: Array<{
+      signing_key: string
+      ledger_index: string
+      unl: boolean
+    }>
+    deprecated: boolean | null
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:54]: When would it be `null`?

File: src/api/routes/v1/amendments.ts [hunk 12]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
+      ledger_index: string
+      unl: boolean
+    }>
+  }
+}
+
+interface AmendmentInVotingMap {
+  [key: string]: {
+    name: string
+    rippled_version: string | undefined | null
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:46]: When would it be null?

File: src/api/routes/v1/amendments.ts [hunk 13]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
+      'amendments_enabled.ledger_index',
+      'amendments_enabled.tx_hash',
+      'amendments_enabled.date',
+      'amendments_info.name',
+      'amendments_info.rippled_version',
+      'amendments_info.deprecated',
+    )
+    .where('amendments_enabled.networks', id)
+
+  enabled.sort((prev: AmendmentsEnabled, next: AmendmentsEnabled) =>
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:152]: This seems like an extra level.  Will `enabled.sort(sortByVersion))` not work?

File: src/api/routes/v1/amendments.ts [hunk 14]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
+  const dbUNL = await query('validators')
+    .count('signing_key AS count')
+    .whereNotNull('unl')
+    .andWhere('chain', network_id)
+
+  const totalUnl: number = dbUNL[0].count
+
+  // eslint-disable-next-line require-atomic-updates -- no race condition.
+  votingMap[amendment_id].threshold = `${Math.ceil(
+    0.8 * totalUnl,
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:219]: Magic number.

File: src/api/routes/v1/amendments.ts [hunk 15]
```diff
@@ -11,45 +12,301 @@ interface AmendmentsInfoResponse {
...
+    )
+    .where('validators.networks', id)
+
+  const votingAmendments: AmendmentInVotingMap = {}
+
+  inNetworks.forEach((val) => {
+    parseAmendmentVote(val, votingAmendments)
+  })
+
+  if (Date.now() - cacheInfo.time > 60 * 1000) {
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:251]: Make this property a constant so that caching time can be tweaked easily.

File: src/connection-manager/data/amendments_enabled.json (hunk 3 above)
Comment (pdp2121) [reply src/connection-manager/data/amendments_enabled.json:312]: This is because of the reset. I think it should not be here in the first place. I will fix on previous PR and merge

File: src/api/routes/v1/amendments.ts (hunk 4 above)
Comment (pdp2121) [reply src/api/routes/v1/amendments.ts:84]: id for networks on VHS are string (either "main", "dev", or "1", "2", etc.)

File: src/api/routes/v1/amendments.ts [hunk 16]
```diff
@@ -43,15 +45,15 @@ interface AmendmentInVoting extends AmendmentsInfo {
 interface AmendmentInVotingMap {
   [key: string]: {
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:46]: Is the key here the amendment hash?

Comment (pdp2121) [reply src/api/routes/v1/amendments.ts:46]: Yes it is. I used map to parse amendments string "id1,id2,id3" into id1: validator1, validator2, etc. more efficiently

File: package.json [hunk 17]
```diff
@@ -55,7 +55,8 @@
     "prettier": "^2.2.1",
     "ts-jest": "^26.4.4",
     "ts-node-dev": "^2.0.0",
-    "typescript": "^4.1.3"
+    "typescript": "^4.1.3",
+    "xrpl": "^2.12.0"
```
Comment (ckniffen) [inline package.json:59]: `xrpl` should not need to be a `devDependency` too.

File: src/api/routes/v1/amendments.ts [hunk 18]
```diff
@@ -1,54 +1,309 @@
...
   count: number
   amendments: AmendmentsInfo[]
 }
 
+interface AmendmentsVoteResponse {
+  result: 'success' | 'error'
+  count: number
+  amendments: Array<AmendmentsEnabled | AmendmentInVoting>
+}
 interface SingleAmendmentInfoResponse {
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:22]: nit: Add new line

File: src/api/routes/v1/amendments.ts [hunk 19]
```diff
@@ -1,54 +1,309 @@
...
   amendments: AmendmentsInfo[]
   time: number
 }
 
+interface AmendmentInVoting extends AmendmentsInfo {
+  threshold: string
+  consensus: string
+  voted: {
+    count: number
+    validators: Array<{
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:37]: This could become its own type.  Also these two types seem like they have a lot of redundancy.

File: src/connection-manager/index.ts [hunk 20]
```diff
@@ -20,7 +19,6 @@ async function start(): Promise<void> {
   }
   await Promise.all(promises)
   await startConnections()
-  await addAmendmentsDataFromJSON()
```
Comment (ckniffen) [inline src/connection-manager/index.ts:23]: Wasnt this just added in the previous PR?  Why was it moved again?

File: src/shared/data/amendments_info.json [hunk 21]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (ckniffen) [inline src/shared/data/amendments_info.json:1]: Why is this empty?

File: src/shared/database/update-amendments-from-json.ts [hunk 22]
```diff
@@ -0,0 +1,98 @@
+import 'dotenv/config'
```
Comment (ckniffen) [inline src/shared/database/update-amendments-from-json.ts:1]: What config properties is this using?

File: src/shared/data/amendments_info.json (hunk 21 above)
Comment (pdp2121) [reply src/shared/data/amendments_info.json:1]: This is to manually add amendment info for amendments that are not on xrpl.org or the rippled files we use (probably ones that are on sidechain). I have this setup first as an empty array but we can add data later on when needed.

File: src/connection-manager/index.ts (hunk 20 above)
Comment (pdp2121) [reply src/connection-manager/index.ts:23]: The one added on the previous PR was `fetchAmendmentInfo` which fetches amendments info from the 2 links. This one fetches amendments data from JSON, and I had it ran on `setupTables` now since it only runs once when we redeploy

**Review by ckniffen — APPROVED** [review]
Please add a note about making a constant for cache time in your commit when you merge

File: src/shared/database/update-amendments-from-json.ts [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+): Promise<void> {
+  await query('amendments_enabled')
+    .insert(enabledAmendment)
+    .onConflict(['amendment_id', 'networks'])
+    .merge()
+    .catch((err) => log.error('Error Saving Enabled Amendment', err))
+}
+
+/**
+ * Add amendments info data from Json file munually to the database.
```
Comment (jonathanlei) [inline src/shared/database/update-amendments-from-json.ts:57]: manually

File: src/shared/database/update-amendments-from-json.ts [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+  log.info('Adding Amendments Information from JSON File...')
+  const data = amendmentInfoData
+  data.forEach(async (amendmentInfo: AmendmentsInfo) => {
+    await saveAmendmentsInfo(amendmentInfo)
+  })
+  log.info('Finished adding Amendments Information from JSON File.')
+}
+
+/**
+ * Add enabled amendments data from Json file munually to the database.
```
Comment (jonathanlei) [inline src/shared/database/update-amendments-from-json.ts:71]: manually

**Review by jonathanlei — APPROVED** [review]
Have some small nits but overall LGTM

File: src/shared/data/amendments_info.json (hunk 21 above)
Comment (pdp2121) [reply src/shared/data/amendments_info.json:1]: This only needed if there's a need so I don't think we need a TODO
